frame/system/src/lib.rs

```diff
@@ -340,6 +341,10 @@ pub mod pallet {
 		/// The block header.
 		type Header: Parameter + traits::Header<Number = Self::BlockNumber, Hash = Self::Hash>;
```

Comment:
You can retrieve the header from the block, no need to have both.

Reply:
I am trying to remove `Header` and add a constraint for `Header::Number` as `Self::BlockNumber`. But, this is resulting in an error: `associated type bounds are unstable`. 

Essentially, I want this:
```rust
type Block: Parameter + traits::Block<Hash = Self::Hash, Header: traits::Header<Number = Self::BlockNumber, Hash = Self::Hash>>;
```

Any suggestions on fixing this?

Reply:
Probably a more radical approach would be to get these types from the `Block` type as well :see_no_evil:

Reply:
I looked into it and it seems that we assume the presence of `BlockNumber` in `frame_system::Config` at a number of places:
1. Hooks in frame/support
2. A number of pallets accessing T::BlockNumber
3. Executive
Please let me know if we are good to make those changes.

Reply:
@bkchr I tried to get the BlockNumber from the Block itself and replaced the same in a number of places. Everything in node-template-runtime seems to compile except the change in calls for `frame/grandpa` which take `T::BlockNumber` as an argument. I realized that this might be because any param in a call needs to implement the `Parameter` trait. So, I moved that trait to sp_runtime, but I still see the same error. 

Do you have some other suggestions that I can try?
Here's the diff but I am not sure it would be very helpful: https://github.com/paritytech/substrate/compare/gupnik/cr1?expand=1